File: projects/rpn2tex/migrations/java-feature-by-feature-3/src/main/java/com/rpn2tex/Parser.java
package com.rpn2tex;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

/**
 * Parser for RPN expressions.
 * <p>
 * The parser uses a stack-based approach to evaluate RPN notation
 * and build an abstract syntax tree (AST). It processes tokens from
 * left to right:
 * </p>
 * <ul>
 *   <li>Numbers are pushed onto the stack</li>
 *   <li>Operators pop operands from the stack and push results</li>
 * </ul>
 * <p>
 * Example usage:
 * </p>
 * <pre>
 * List&lt;Token&gt; tokens = lexer.tokenize();
 * Parser parser = new Parser(tokens);
 * Expr ast = parser.parse();
 * </pre>
 */
public class Parser {
    private final List<Token> tokens;
    private int position;

    /**
     * Constructs a new Parser for the given token list.
     *
     * @param tokens The list of tokens to parse
     */
    public Parser(List<Token> tokens) {
        this.tokens = tokens;
        this.position = 0;
    }

    /**
     * Parses the tokens and returns the root AST node.
     *
     * @return The root expression node
     * @throws ParserException if the token sequence is invalid
     */
    public Expr parse() throws ParserException {
        Deque<Expr> stack = new ArrayDeque<>();

        while (!atEnd()) {
            Token token = current();

            if (token.type() == TokenType.EOF) {
                break;
            }

            if (token.type() == TokenType.NUMBER) {
                // Push number onto stack
                NumberExpr numberExpr = new NumberExpr(
                        token.value(),
                        token.line(),
                        token.column()
                );
                stack.push(numberExpr);
                advance();
            } else if (token.type() == TokenType.PLUS || token.type() == TokenType.MINUS || token.type() == TokenType.TIMES || token.type() == TokenType.DIVIDE) {
                // Binary operator: pop two operands, push result
                if (stack.size() < 2) {
                    throw new ParserException(
                            "Operator '" + token.value() + "' requires two operands",
                            token
                    );
                }

                Expr right = stack.pop();
                Expr left = stack.pop();

                // Map token type to operator string
                String operator;
                if (token.type() == TokenType.PLUS) {
                    operator = "+";
                } else if (token.type() == TokenType.MINUS) {
                    operator = "-";
                } else if (token.type() == TokenType.TIMES) {
                    operator = "*";
                } else {
                    operator = "/";
                }

                BinaryOpExpr binaryOpExpr = new BinaryOpExpr(
                        operator,
                        left,
                        right,
                        token.line(),
                        token.column()
                );
                stack.push(binaryOpExpr);
                advance();
            } else {
                // Unknown token type
                throw new ParserException("Unexpected token: " + token.type(), token);
            }
        }

        // Validate stack state
        if (stack.isEmpty()) {
            Token eofToken = tokens.get(tokens.size() - 1);
            throw new ParserException("Empty expression", eofToken);
        }

        if (stack.size() > 1) {
            throw new ParserException(
                    "Expression has too many operands (expected 1, got " + stack.size() + ")",
                    tokens.get(tokens.size() - 1)
            );
        }

        return stack.pop();
    }

    /**
     * Returns the current token without consuming it.
     *
     * @return The current token
     */
    private Token current() {
        if (atEnd()) {
            return tokens.get(tokens.size() - 1); // Should be EOF
        }
        return tokens.get(position);
    }

    /**
     * Advances to the next token.
     */
    private void advance() {
        if (!atEnd()) {
            position++;
        }
    }

    /**
     * Checks if we've reached the end of the token list.
     *
     * @return true if at end, false otherwise
     */
    private boolean atEnd() {
        return position >= tokens.size() || tokens.get(position).type() == TokenType.EOF;
    }
}

File: projects/rpn2tex/migrations/go-feature-by-feature-2/parser.go
package rpn2tex

// Parser converts a token stream into an AST.
type Parser struct {
	tokens []Token
	pos    int
}

// NewParser creates a new Parser for the given token stream.
func NewParser(tokens []Token) *Parser {
	return &Parser{
		tokens: tokens,
		pos:    0,
	}
}

// Parse parses the token stream and returns the root AST node.
func (p *Parser) Parse() (Expr, error) {
	var stack []Expr

	for !p.atEnd() {
		token := p.current()

		if token.Type == TokenNumber {
			node := &Number{
				Value:  token.Value,
				Line:   token.Line,
				Column: token.Column,
			}
			stack = append(stack, node)
			p.advance()
		} else if token.Type == TokenPlus || token.Type == TokenMinus || token.Type == TokenMult || token.Type == TokenDiv {
			// Binary operator: pop 2 operands
			opSymbol := "+"
			if token.Type == TokenMinus {
				opSymbol = "-"
			} else if token.Type == TokenMult {
				opSymbol = "*"
			} else if token.Type == TokenDiv {
				opSymbol = "/"
			}

			if len(stack) < 2 {
				return nil, &ParserError{
					Message: "Operator '" + opSymbol + "' requires two operands",
					Token:   token,
				}
			}
			// Pop right, then left (stack order)
			right := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			left := stack[len(stack)-1]
			stack = stack[:len(stack)-1]

			node := &BinaryOp{
				Operator: opSymbol,
				Left:     left,
				Right:    right,
				Line:     token.Line,
				Column:   token.Column,
			}
			stack = append(stack, node)
			p.advance()
		} else if token.Type == TokenEOF {
			break
		} else {
			return nil, &ParserError{
				Message: "Unexpected token type",
				Token:   token,
			}
		}
	}

	if len(stack) == 0 {
		return nil, &ParserError{
			Message: "Empty expression",
			Token:   Token{Line: 1, Column: 1},
		}
	}

	if len(stack) > 1 {
		return nil, &ParserError{
			Message: "Too many operands",
			Token:   p.tokens[0],
		}
	}

	return stack[0], nil
}

// current returns the current token without consuming it.
func (p *Parser) current() Token {
	if p.atEnd() {
		return p.tokens[len(p.tokens)-1] // Return EOF token
	}
	return p.tokens[p.pos]
}

// advance moves to the next token.
func (p *Parser) advance() {
	if !p.atEnd() {
		p.pos++
	}
}

// atEnd checks if we've reached the end of the token stream.
func (p *Parser) atEnd() bool {
	return p.pos >= len(p.tokens) || p.tokens[p.pos].Type == TokenEOF
}
